XOR

mrwatts88 opened the pull request:
Add the ability to evaluate rules composed with XOR, and the helper function, similar to and, or, not.

mrwatts88 committed:
add the ability to evaluate rules composed with XOR, and the helper function, similar to and, or, not.

eisenivan commented on src/impl.spec.js:
```diff
@@ -20,6 +20,10 @@ test('Implement Regent without init()', (assert) => {
 
   assert.true(evaluate(SHOULD_WEAR_COAT, data), 'Should Wear Coat should return true');
+
+  const WET_BUT_NOT_TOO_WET = xor(IS_RAINING, IS_SNOWING);
```
Love the example!

eisenivan commented on src/index.js:
```diff
@@ -25,6 +25,11 @@ export const or = (...rules) => ({
   rules,
 });
 
+export const xor = (...rules) => ({
```
We will probably want a test or two making sure that this function exists, and that checks it's param signature. So something like 

```
assert.equal(typeof xor, 'function')
```

eisenivan commented on src/private/parse-composed.js:
```diff
@@ -9,17 +9,21 @@ export default (obj, data, custom = {}) => {
     result = !evaluateRule(notRule, data, custom);
   } else {
     const action = obj.compose;
+    const fxn = rule => (
```
This is great too! Nice abstraction.

eisenivan reviewed (changes requested):
Thanks so much for the PR. We're taking a look at it and will be in touch, but in the mean time I left a few notes.

Could you add unit tests for `index.js`, and `parse-composed.js`? There are `.spec.js` files as siblings to both with a bunch of examples.

How did you stumble on regent? Hope you don't mind me asking.

mrwatts88 commented:
I will look into those changes today. I was browsing the Milwaukee Slack channel and I'll be interning at NWM this Summer so Regent caught my eye.

eisenivan commented:
Awesome! We are also considering caping the number of arguments `xor` gets to 2. We were initially thinking that it should take `n` arguments, and confirm that some but not all are true, like you implemented, but now we think it would be more clear if it took exactly 2 and threw an error if you passed it more.

Do you have any thoughts on that? I'd love to get some outside thinking on it.

This Wikipedia entry sort of illustrates our thinking. Thanks to @aLittleTooQuiet for digging that up.

https://en.wikipedia.org/wiki/XOR_gate#More_than_two_inputs

mrwatts88 committed:
Add XOR tests

Add tests for XOR functionality in index.spec.js and parse-composed.spec.js

mrwatts88 commented:
It's hard to say, but the way I've implemented it isn't typically the way XOR works with multiple inputs. In a digital logic sense, it typically outputs true if the number of true inputs is odd, false otherwise. So there are really 2 options for how to implement XOR with 3 or more arguments. Restricting the number of args to 2 will eliminate that ambiguity.

Tough call, but anyways I have added the tests you asked for. Let me know if you want to make the XOR change.

mrwatts88 committed:
Limit XOR arguments to exactly 2 rules.

Throw Error when creating a composed xor object with any number of arguments != 2, also throw Error when calling evaluate with a manually constructed composed xor object with number of rules !=2. Add tests for new implementation.

eisenivan commented:
@mrwatts88 this looks great! Only tests I'd like to see added would be sad path tests, so basically ensuring it throws an error when called with 0 rules, 1 rule, 3+ rules.

mrwatts88 committed:
Add XOR sad path tests

Add tests for xor (using helper method) in impl.spec.js, and in evaluate-rule.spec.js. xor should throw with 0, 1 or more than 2 arguments.

mrwatts88 committed:
fix block padding

mrwatts88 commented:
@eisenivan okay I've added those tests. Let me know if they aren't in the correct place or if there is anything you need me to change.
